### PR TITLE
feat(agw): add sctp up/downstream sock to mme service config

### LINKED ISSUE #8828: [AGW][vN] MME build flag to enable SCTPD proxy via accessd
Our preference here is to enable over-ride of the Unix Domain Socket for uplink / downlink SCTP traffic through sctpd.

The configuration will take place in the `MME`, allowing it to optionally be redirected away from the standard SCTP uplink / downlink Unix Domain Sockets offered by `sctpd`, and instead to the set offered by `accessd` - the golang proxy.

Configuration based change (e.g. through Magma's `mconfigs`) is preferred, though a gflag fallback could be acceptable.

### PR DESCRIPTION
**Pull request by electronjoe**
## Summary

Closes #8828.

Informed by https://github.com/magma/magma/commit/7c54225801bcd75a8d4ab6c9bf4190100745372c

### Changes

- Removed unused sctp configurations from MME service config
  - in_streams (SCTP_INSTREAMS in config)
  - out_streams (SCTP_OUTSTREAMS in config)
- Adds sctp configurations for MME service config
  - SCTP_UPSTREAM_SOCK (default unix:///tmp/sctpd_upstream.sock)
  - SCTP_DOWNSTREAM_SOCK (default unix:///tmp/sctpd_downstream.sock)

## Test Plan

- Ran `lte integration test` to ensure no regression in nominal parameter
- Added unit test to verify configuration parsing of new fields
- In magma VM, ensured config change results in new unix domain socket path
  - vagrant ssh magma
  - cd `/magma/lte/gateway`
  - vim `/vagrant/configs/templates/mme.conf.template`
    - update `SCTP_UPSTREAM_SOCK = unix:///tmp/sctpd_upstream_test.sock`
    - update `SCTP_DOWNSTREAM_SOCK = unix:///tmp/sctpd_downstream_test.sock`
  - make run
  - vim `/var/opt/magma/tmp/mme.conf` (validate config)
    - validate that the template was filled correctly (should be no-op for new parameters)
  - ls -lh `/tmp/sctpd_*.sock`
    - should contain
      - sctpd_downstream.sock (generated by sctpd)
      - /tmp/sctpd_upstream_2.sock (generated by mme with new config)

Signed-off-by: Scott Harrison Moeller <smoeller@fb.com>

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/mme_app/mme_config.c [hunk 1]
```diff
@@ -1887,11 +1907,11 @@ void mme_config_display(mme_config_t* config_pP) {
       bdata(config_pP->itti_config.log_file));
   OAILOG_INFO(LOG_CONFIG, "- SCTP:\n");
   OAILOG_INFO(
-      LOG_CONFIG, "    in streams .......: %u\n",
-      config_pP->sctp_config.in_streams);
+      LOG_CONFIG, "  upstream_sctp_sock..: %s\n",
+      bdata(config_pP->sctp_config.upstream_sctp_sock));
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_config.c:2015]: we should free the according added bstrings on `free_mme_config` function

**Review by ardzoht — CHANGES_REQUESTED** [review]
requesting changes to add free calls to the up/downstream sock bstring

File: lte/gateway/c/core/oai/tasks/mme_app/mme_config.c (hunk 1 above)
Comment (electronjoe) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_config.c:2015]: Seems I did not push my most recent branchy - this issue was actually detected by my unit test when I added it (wooho unit tests + ASAN). Pushing the latest branch which does as you suggest (nice review!)

**Review by ardzoht — APPROVED** [review]
lgtm

**Comment by electronjoe** [comment]
@rdefosse this is failing Magma-OAI-Jenkins' `ds_tester` phase. Trying to assess root cause or flake?

Any insights from [this log](https://gist.github.com/electronjoe/30f05c095d5d5d9e0368b428dd65d138)

**Comment by rdefosse** [comment]
@electronjoe 

Looks like the `scptd` listener is not OK with the eNB doing a `S1AP` request. In both the `DsTester` sub-pipeline and in the `RHEL8` sanity check.

in `https://jenkins-oai.eurecom.fr/job/MAGMA-MME-RHEL8-Deploy-Check/1018/` --> `archives/sctpd.log`

```bash
I0924 13:51:35.843222    75 sctp_connection.cpp:89] starting sctp connection listener sd = 15
I0924 13:52:33.645169    75 sctp_connection.cpp:167] HandleClientSock sd = 17
I0924 13:52:33.645215    75 sctp_connection.cpp:189] SCTP association change event received
I0924 13:52:33.645546    75 sctpd_uplink_client.cpp:56] sctpul.newassoc error
I0924 13:52:33.645556    75 sctpd_uplink_client.cpp:57] grpc error (14): Connect Failed
```

and in `https://jenkins-oai.eurecom.fr/job/MAGMA-MME-OAI-DsTester/1743/` it is less obvious.

File: lte/gateway/c/core/oai/tasks/sctp/sctpd_downlink_client.cpp [hunk 2]
```diff
@@ -110,8 +109,10 @@ constexpr useconds_t max_backoff_usecs = 1000000;  // 1 sec
 std::unique_ptr<SctpdDownlinkClient> client = nullptr;
 
 int init_sctpd_downlink_client(bool force_restart) {
-  auto channel =
-      grpc::CreateChannel(DOWNSTREAM_SOCK, grpc::InsecureChannelCredentials());
+  std::string downstream_sctp_sock(
+      bstr2cstr(mme_config.sctp_config.downstream_sctp_sock, '\0'));
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/sctp/sctpd_downlink_client.cpp:113]: this leaks memory as bstr2cstr will allocate a new memory while string object will have its own copy of it. we should use `bdata(mme_config.sctp_config.downstream_sctp_sock)` instead.

File: lte/gateway/c/core/oai/tasks/sctp/sctpd_uplink_server.cpp [hunk 3]
```diff
@@ -140,7 +140,10 @@ int start_sctpd_uplink_server(void) {
   service = std::make_shared<SctpdUplinkImpl>();
 
   ServerBuilder builder;
-  builder.AddListeningPort(UPSTREAM_SOCK, grpc::InsecureServerCredentials());
+  std::string upstream_sctp_sock(
+      bstr2cstr(mme_config.sctp_config.upstream_sctp_sock, '\0'));
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/sctp/sctpd_uplink_server.cpp:144]: same issue. we should use `bdata` here instead of `bstr2cstr`.

**Comment by ulaskozat** [comment]
for the memory leak issue: https://github.com/magma/magma/issues/9479

**Pull request merged**